src/t_string.c

```diff
@@ -90,73 +94,143 @@ void setGenericCommand(client *c, int flags, robj *key, robj *val, robj *expire,
     }
 
     if (flags & OBJ_SET_GET) {
-        if (getGenericCommand(c) == C_ERR) return;
+        if (getGenericCommand(c, OBJ_NO_FLAGS, NULL, 0) == C_ERR) return;
     }
 
-    genericSetKey(c,c->db,key,val,flags & OBJ_SET_KEEPTTL,1);
+    genericSetKey(c,c->db,key, val,flags & OBJ_KEEPTTL,1);
     server.dirty++;
-    if (expire) setExpire(c,c->db,key,mstime()+milliseconds);
+    if (expire && ((flags & OBJ_PX) || (flags & OBJ_EX))) {
+        milliseconds += mstime();
+    }
     notifyKeyspaceEvent(NOTIFY_STRING,"set",key,c->db->id);
-    if (expire) notifyKeyspaceEvent(NOTIFY_GENERIC,
-        "expire",key,c->db->id);
+    if (expire) {
+        setExpire(c, c->db, key, milliseconds);
+        notifyKeyspaceEvent(NOTIFY_GENERIC,
+                            "expire", key, c->db->id);
+
+        // Propagate as SET Key Value PXAT millisecond-timestamp
+        robj* millisecondObj = createStringObjectFromLongLong(milliseconds);
+        rewriteClientCommandVector(c, 5, shared.set, key, val, shared.pxat, millisecondObj);
+        decrRefCount(millisecondObj);
```

Comment:
Thanks @oranagra for the feedback. Does the recommendation also apply if we get `PXAT/EXAT`, Do we want to convert absolute TTL to relative in this case?

Reply:
no, if we're asked for absolute time, we forward that intent to the replica as is.
AFAIK that's what the code does, and AFAIR from that conversation with Salvatore, that's what he intended.

Reply:
Thank you! @oranagra for the review. I will handle this and rest of the comments soon.

Reply:
i'm trying to add a test for that for the existing commands (with a reference to the justification), so you can skip that for now, and just work on fixing the changes in this PR.

Reply:
ACK thanks!

Reply:
https://github.com/redis/redis/pull/8357

Reply:
@oranagra Can we re-open that conversation? It seems wrong to have behavior that might change dramatically on failovers if the nodes have different clocks. I don't really buy salvatore's argument we should only care about the typical case. Large replication lag, full syncs, psyncs can all introduce a lot of lag between when an expire is set and when it's propagated to a replica. Clock skew is usually on the matter of seconds, but these modes could introduce error on the order of minutes/hours.

I see shootIT commented on your test PR as well.

Reply:
@madolson sure we can open that for conversation, i just want it to be intentional rather than overlooked.
i suggest to have that conversation separate from this PR.
It can be in my PR (before or after merging it), or in another issue.

Reply:
I'll re-open an issue.